docs: Operator Go Versioning Policy (#3010)
redhat-developer/rhdh-operator: docs/go-version-policy.md

Reviewer: Consider adding a quick TL;DR or concise summary table at the top as this is a bit long IMO. This will help maintainers quickly find the right action without having to read 110+ lines ;)
The summary is actually highlighted in https://redhat.atlassian.net/browse/RHIDP-14096 (`Policy to document` section).

Reviewer 2: +1
This Policy contains nice description and motivation pieces but in most cases we need a quick, practical guide to timely version updates.

# Go Versioning Policy for the RHDH Operator

## TL;DR

| Control | Main branch | Release branches |
|---|---|---|
| `go` directive in `go.mod` | Bump to match the latest Go version available in go-toolset | Frozen at branch-cut value; do not bump |

> | `toolchain` directive in `go.mod` | Update to the latest patch release of the declared `go` version | Frozen at branch-cut value; do not bump |

Reviewer: Well, even on release branches, it should be fine to receive patch-level updates to the toolchain (as long as the matching declared `go` version is still supported). If the declared `go` (frozen in release branches) is no longer supported, then we will stop receiving such toolchain updates. I think Renovate does that (stated in L67).

> | go-toolset builder image | Track the latest available image | Track the latest available image (security patches) |

Reviewer: ```suggestion
| go-toolset builder image | Track the latest available image | Track the latest available image |
```

Why this explicit statement here? Might be confusing IMO.

## Language Version Policy on the Main Branch

The `go` directive on the main branch is eventually bumped to match the latest Go version available in the Red Hat [go-toolset](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690) image, which in practice can lag behind the latest upstream Go release. This lag is not a policy choice but a consequence of the go-toolset image availability: the Red Hat go-toolset typically may lag behind official upstream Go releases by several months while the image is rebuilt and certified against the current RHEL base. See https://access.redhat.com/articles/7116095.

A new Go version is considered available for the operator when a go-toolset image containing that version appears in the [Red Hat Ecosystem Catalog](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690). Maintainers can monitor availability by checking the catalog for new tags or by observing when Renovate opens a Docker minor update pull request for the go-toolset image in the operator repository. Updates to the `go` directive are driven by this availability, not by upstream Go release dates.

> When the `go` directive is bumped, the `toolchain` directive should be updated to track the latest patch release of the declared version. Both changes are deliberate and tracked via a Jira issue each release cycle (for example, [RHIDP-12020](https://redhat.atlassian.net/browse/RHIDP-12020) tracks the update to Go 1.26). At the same time, the [`constraints.go` setting](https://docs.renovatebot.com/golang/#go-binary-version) in the Renovate configuration at [`.github/renovate.json`](.github/renovate.json) must be updated to match the new `go` directive value. This constraint controls which Go dependency versions Renovate is permitted to propose, and a mismatch will cause Renovate to offer updates that may be incompatible with the declared language version.

Reviewer: ```suggestion
When the `go` directive is bumped, the `toolchain` directive should be updated to track the latest patch release of the declared version. Both changes are deliberate and tracked via a Jira issue each release cycle (for example, [RHIDP-12020](https://redhat.atlassian.net/browse/RHIDP-12020) tracks the update to Go 1.26). At the same time, the [`constraints.go` setting](https://docs.renovatebot.com/golang/#go-binary-version) in the Renovate configuration at [`.github/renovate.json`](../.github/renovate.json) must be updated to match the new `go` directive value. This constraint controls which Go dependency versions Renovate is permitted to propose, and a mismatch will cause Renovate to offer updates that may be incompatible with the declared language version.
```

The `go` and `toolchain` directive bumps in `go.mod` are tracked and executed separately from go-toolset image updates. They are deliberate changes managed through the Jira release cycle process, not automated by Renovate. Renovate does automate patch-level updates to the `toolchain` directive (for example, `go1.25.9` to `go1.25.10`) and opens pull requests for these changes. These pull requests require manual review and are not automerged, which gives maintainers an opportunity to verify the patch before it reaches a release branch.

## Compatibility Guarantee

This policy relies on Go's strong backward compatibility guarantees. The [Go 1 compatibility promise](https://go.dev/doc/go1compat) states that `it is intended that programs written to the Go 1 specification will continue to compile and run correctly, unchanged, over the lifetime of that specification.`.

> The [GODEBUG mechanism](https://go.dev/doc/godebug), described in the Release Branches section above, provides an additional safety layer by gating backward-incompatible changes behind settings derived from the `go` directive. The nightly test suite provides further coverage against compatibility regressions. If a go-toolset update introduces a behavioral change that affects the operator despite the compatibility promise and GODEBUG protection, the nightly tests are expected to surface the issue before it reaches a release.

Reviewer: ```suggestion
The [GODEBUG mechanism](https://go.dev/doc/godebug), described in the Release Branches section above, provides an additional safety layer. The nightly test suite provides further coverage against compatibility regressions. If a go-toolset update introduces a behavioral change that affects the operator despite the compatibility promise and GODEBUG protection, the nightly tests are expected to surface the issue before it reaches a release.
```

Already explained above.

**[Universal Base Image (UBI)](https://access.redhat.com/articles/4238681)** is Red Hat's freely redistributable, OCI-compliant container base image built from Red Hat Enterprise Linux. UBI images are available without a subscription from the unauthenticated registry at `registry.access.redhat.com` and from the authenticated registry at `registry.redhat.io`. All container images shipped as part of a Red Hat product must be built on UBI to meet [Red Hat's container certification requirements](https://developers.redhat.com/articles/ubi-faq). The operator uses two UBI variants: `ubi9/go-toolset` as the builder stage and [`ubi9-minimal`](https://catalog.redhat.com/en/software/containers/ubi9/ubi-minimal/61832888c0d15aff4912fe0d) as the runtime stage. The UBI content availability and life cycle are governed by the [corresponding RHEL release](https://access.redhat.com/support/policy/updates/ubi).

## Language Version Policy on the Main Branch

> The `go` directive on the main branch tracks the latest Go version available in the Red Hat [go-toolset](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690) image, which in practice tends to be one major version behind the latest upstream Go release. This lag is not a policy choice but a consequence of the go-toolset image availability: the Red Hat go-toolset typically lags behind official upstream Go releases by several months while the image is rebuilt and certified against the current RHEL base. Go's [release policy](https://go.dev/doc/devel/release) supports the two most recent major versions, so even with this lag the module remains within the supported window.

Reviewer: ```suggestion
The `go` directive on the main branch is eventually bumped to match the latest Go version available in the Red Hat [go-toolset](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690) image, which in practice can lag behind the latest upstream Go release. This lag is not a policy choice but a consequence of the go-toolset image availability: the Red Hat go-toolset typically may lag behind official upstream Go releases by several months while the image is rebuilt and certified against the current RHEL base. See https://access.redhat.com/articles/7116095.
```

`track` implies they always stay in sync, but this may not be true depending on when you look at the main branch (until the actual bump task is implemented).
Also, `one major version behind the latest upstream Go release.` is misleading. Currently both toolset and upstream Go are at 1.26, but the toolset image may be just one or multiple patches behind. So this strong statement may not always hold.
Also, we are actually bound by the Red Hat support window of Go. To clarify, I think even a Go version that is no longer supported upstream may continue to be supported by RH (with appropriate backports). See https://access.redhat.com/articles/7116095

A new Go version is considered available for the operator when a go-toolset image containing that version appears in the [Red Hat Ecosystem Catalog](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690). Maintainers can monitor availability by checking the catalog for new tags or by observing when Renovate opens a Docker minor update pull request for the go-toolset image in the operator repository. Updates to the `go` directive are driven by this availability, not by upstream Go release dates.

> When the `go` directive is bumped, the `toolchain` directive should be updated to track the latest patch release of the declared version. Both changes are deliberate and tracked via a Jira issue each release cycle (for example, [RHIDP-12020](https://redhat.atlassian.net/browse/RHIDP-12020) tracks the update to Go 1.26). At the same time, the `constraints.go` setting in the Renovate configuration at `.github/renovate.json` must be updated to match the new `go` directive value. This constraint controls which Go dependency versions Renovate is permitted to propose, and a mismatch will cause Renovate to offer updates that are incompatible with the declared language version.

Reviewer: ```suggestion
When the `go` directive is bumped, the `toolchain` directive should be updated to track the latest patch release of the declared version. Both changes are deliberate and tracked via a Jira issue each release cycle (for example, [RHIDP-12020](https://redhat.atlassian.net/browse/RHIDP-12020) tracks the update to Go 1.26). At the same time, the [`constraints.go` setting](https://docs.renovatebot.com/golang/#go-binary-version) in the Renovate configuration at [`.github/renovate.json`](.github/renovate.json) must be updated to match the new `go` directive value. This constraint controls which Go dependency versions Renovate is permitted to propose, and a mismatch will cause Renovate to offer updates that may be incompatible with the declared language version.
```

This is safe because the Go version declared in `go.mod` does not determine the standard library linked into the shipped binary. The binary's standard library comes from the go-toolset compiler used at build time. As long as the go-toolset image is kept current (see the next section), the shipped binary continues to receive Go standard library and compiler security patches regardless of what `go.mod` declares.

The [GODEBUG mechanism](https://go.dev/doc/godebug) ensures that a newer compiler building a module with an older `go` directive will respect the behavior semantics of the declared version. When a Go release introduces a backward-incompatible change, the new behavior is gated behind a GODEBUG setting whose default is derived from the `go` directive in `go.mod`. A binary compiled with Go 1.26 against a module declaring `go 1.24` will therefore use the Go 1.24 behavior for any GODEBUG-controlled changes, preserving the expected runtime semantics without requiring a directive bump.

> ## Build Image Policy

Reviewer: The "Definitions" section already defines UBI and go-toolset but then re-explains them later in "Build Image Policy." Some of this is intentional reinforcement, but it adds length. Could be tightened.

The operator's Dockerfile references the go-toolset image from the unauthenticated registry at `registry.access.redhat.com/ubi9/go-toolset`. The `#@follow_tag` annotation in the Dockerfile indicates the corresponding authenticated registry path at `registry.redhat.io/rhel9/go-toolset` and is used by automated tooling to track the latest available tag. The image reference is pinned by digest for reproducibility; the digest is updated when a new go-toolset image becomes available.

The runtime stage of the Dockerfile uses [`ubi9-minimal`](https://catalog.redhat.com/en/software/containers/ubi9/ubi-minimal/61832888c0d15aff4912fe0d), a stripped-down UBI variant that includes only `microdnf` and a minimal set of packages. The compiled operator binary is copied from the builder stage into this minimal image, keeping the final shipped image small and reducing the attack surface.

> The version gap between the `go.mod` directive and the go-toolset compiler on release branches is expected and intentional. It is a direct consequence of freezing `go.mod` while continuing to update the build image, and it is the mechanism that allows release branches to remain secure without destabilizing the module's declared compatibility.

Reviewer: > The version gap between the `go.mod` directive and the go-toolset compiler on release branches is expected and intentional.

it should be noted that this version gap may also exist on `main` for some time, until the actual bump task is implemented. For example currently we have go 1.25 vs go-toolset Go 1.26 on main.

Docker base image updates to the go-toolset image are classified by Renovate as minor updates when the RHEL version component of the tag changes (for example, `9.7` to `9.8`). However, a Docker minor update can carry a major Go version change inside the image. The go-toolset `9.7` image shipped Go 1.25, while go-toolset `9.8` ships Go 1.26. The pull request diff will only show the image tag and digest change; it will not indicate the Go version change. Reviewers must check the Go version inside the image before approving, using the `skopeo inspect` command described in the Verification section below.

Docker minor updates to go-toolset are not automerged by the Renovate configuration. They produce a pull request that requires manual review and approval. This is intentional and must not be changed, because these updates may introduce a new Go compiler version that should be verified against the nightly test suite before merging.

> Go dependency updates are governed by the `constraints.go` setting in the Renovate configuration. This setting restricts the Go versions that Renovate will consider when evaluating whether a dependency update is compatible. The value must match the `go` directive in `go.mod` on each branch. When the `go` directive is bumped on the main branch, the `constraints.go` value must be updated in the same change or immediately after. A stale `constraints.go` value will either block valid updates or allow incompatible ones.

Reviewer: ```suggestion
Go dependency updates are governed by the [`constraints.go`](https://docs.renovatebot.com/golang/#go-binary-version) setting in the Renovate configuration. This setting restricts the Go versions that Renovate will consider when evaluating whether a dependency update is compatible. The value must match the `go` directive in `go.mod` on each branch. When the `go` directive is bumped on the main branch, the `constraints.go` value must be updated in the same change or immediately after. A stale `constraints.go` value will either block valid updates or allow incompatible ones.
```

The `go` and `toolchain` directive bumps in `go.mod` are tracked and executed separately from go-toolset image updates. They are deliberate changes managed through the Jira release cycle process, not automated by Renovate. Renovate does automate patch-level updates to the `toolchain` directive (for example, `go1.25.9` to `go1.25.10`) and opens pull requests for these changes. These pull requests require manual review and are not automerged, which gives maintainers an opportunity to verify the patch before it reaches a release branch.

## Compatibility Guarantee

> This policy relies on Go's strong backward compatibility guarantees. The [Go 1 compatibility promise](https://go.dev/doc/go1compat) states that programs written to the Go 1 specification will continue to compile and run correctly across future Go 1.x releases. Each release note since Go 1.0 has reiterated this promise, typically with the statement: "As always, the release maintains the Go 1 promise of compatibility. We expect almost all Go programs to continue to compile and run as before."

Reviewer: ```suggestion
This policy relies on Go's strong backward compatibility guarantees. The [Go 1 compatibility promise](https://go.dev/doc/go1compat) states that `it is intended that programs written to the Go 1 specification will continue to compile and run correctly, unchanged, over the lifetime of that specification.`. Each release note since Go 1.0 has reiterated this promise, typically with the statement: "As always, the release maintains the Go 1 promise of compatibility. We expect almost all Go programs to continue to compile and run as before."
```

Where does the last quoted statement come from?

Author: That quote doesn't come from any specific Go documentation page it's a paraphrase of language that appears in various Go release notes (e.g., the Go 1.21, 1.22, 1.23 release notes all have similar wording), but it's not an exact quote from any single source. we can just link to docs here

> The [GODEBUG mechanism](https://go.dev/doc/godebug) provides an additional safety layer. When a Go release introduces a change that could affect existing programs, the change is gated behind a GODEBUG setting. The default value of that setting is determined by the `go` directive in the main module's `go.mod`. This means that building with a Go 1.26 compiler while declaring `go 1.24` in `go.mod` will produce a binary that behaves like Go 1.24 for any GODEBUG-controlled changes, while still receiving the compiler and standard library security improvements from Go 1.26.

Reviewer: I feel like this is over-explained. It seems to be covered in 3 separate sections (Definisions, Release Branches, Compatibility Guarantee) with a lot of repetition. I'd suggest keeping the explanation maybe in the Release Branches section only, as that's where the reader needs to understand why freezing is safe.

## Verification

To check the Go compiler version inside a go-toolset image tag before approving a Renovate pull request, run:

```
skopeo inspect docker://registry.access.redhat.com/ubi9/go-toolset:<tag> | \
  python3 -c "import sys,json; print(json.load(sys.stdin)['Labels']['version'])"
```

Replace `<tag>` with the tag from the pull request (for example, `9.8-1777889793`).

To confirm the Go version linked into a shipped operator binary, inspect the software bill of materials:

```
cosign download sbom --platform=linux/amd64 quay.io/rhdh/rhdh-rhel9-operator:<tag> 2>/dev/null | \
  jq -r '.packages[] | select(.name == "stdlib") | .versionInfo'
```

Replace `<tag>` with the operator image tag (for example, `1.8` or `1.8-91`).

> To check available go-toolset versions and determine when a new Go version has become available for the operator, browse the [Go Toolset for UBI 9](https://catalog.redhat.com/en/software/containers/ubi9/go-toolset/61e5c00b4ec9945c18787690) page in the Red Hat Ecosystem Catalog. The catalog lists all available tags and their associated Go versions. Alternatively, query the registry directly:

Reviewer: At some point, we may need to switch to UBI 10. So I would suggest not hardcoding the UBI version in this document.
